chainntnfs/txconfnotifier.go

```diff
@@ -108,49 +111,105 @@ func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32) *TxConfNotif
 
 // Register handles a new notification request. The client will be notified when
 // the transaction gets a sufficient number of confirmations on the blockchain.
-// If the transaction has already been included in a block on the chain, the
-// confirmation details must be given as the txConf argument, otherwise it
-// should be nil. If the transaction already has the sufficient number of
-// confirmations, this dispatches the notification immediately.
-func (tcn *TxConfNotifier) Register(ntfn *ConfNtfn, txConf *TxConfirmation) error {
+//
+// NOTE: If the transaction has already been included in a block on the chain,
+// the confirmation details must be provided with the UpdateConfDetails method,
+// otherwise we will wait for the transaction to confirm even though it already
+// has.
+func (tcn *TxConfNotifier) Register(ntfn *ConfNtfn) error {
 	select {
 	case <-tcn.quit:
-		return fmt.Errorf("TxConfNotifier is exiting")
+		return ErrTxConfNotifierExiting
 	default:
 	}
 
-	if txConf == nil || txConf.BlockHeight > tcn.currentHeight {
-		// Transaction is unconfirmed.
-		tcn.confNotifications[*ntfn.TxID] =
-			append(tcn.confNotifications[*ntfn.TxID], ntfn)
+	tcn.Lock()
+	defer tcn.Unlock()
+
+	ntfns, ok := tcn.confNotifications[*ntfn.TxID]
+	if !ok {
+		ntfns = make(map[uint64]*ConfNtfn)
+		tcn.confNotifications[*ntfn.TxID] = ntfns
+	}
+
+	ntfns[ntfn.ConfID] = ntfn
+
+	return nil
+}
+
+// UpdateConfDetails attempts to update the confirmation details for the an
+// active notification within the notifier. This should only be used in the case
+// of a transaction that has confirmed before the notifier's current height.
+//
+// NOTE: The notification should be registered first to ensure notifications are
+// dispatched correctly.
+func (tcn *TxConfNotifier) UpdateConfDetails(txid chainhash.Hash, clientID uint64,
+	details *TxConfirmation) error {
+
+	select {
+	case <-tcn.quit:
+		return ErrTxConfNotifierExiting
+	default:
+	}
+
+	// Ensure we hold the lock throughout handling the notification to
+	// prevent the notifier from advancing its height underneath us.
+	tcn.Lock()
+	defer tcn.Unlock()
+
+	// First, we'll determine whether we have an active notification for
+	// this transaction with the given ID.
+	ntfns, ok := tcn.confNotifications[txid]
+	if !ok {
+		return fmt.Errorf("no notifications found for txid %v", txid)
+	}
+
+	ntfn, ok := ntfns[clientID]
+	if !ok {
+		return fmt.Errorf("no notification found with ID %v", clientID)
+	}
+
+	// If the notification has already recognized that the transaction
+	// confirmed, there's nothing left for us to do.
+	if ntfn.details != nil {
+		return nil
+	}
+
+	// The notifier has yet to reach the height at which the transaction was
```

Comment:
This is a bit confusing, how would we _ever_ know in advance if the conf intent is already satisfied? Seems like it should read `DispatchBlockHeight` here instead of just a block height.

Reply:
If it's the block height that the transaction was _initially_ confirmed at, why can we determine this independent of the notifier itself advancing? Do we have a test to exercise this case?

Reply:
The reason behind the check is due to the lag between the `TxConfNotifier` and the backend. For example, let's say that the notifier starts at height 100 and the the transaction confirms at 102. The backend moves up to 102, but the `TxConfNotifier` is still processing height 101. In the event that `UpdateConfDetails` is called before `ConnectTip(..., 102, ...)`, the notifier isn't yet aware of this height, so we defer until handling the notification dispatch until then.

Not sure what you mean by `DispatchBlockHeight`. Are you referring to renaming the method `UpdateConfDetails` to that?